### PR TITLE
[4.0] Prepared plg system remember

### PR DESCRIPTION
**Pull request by HLeithner**
# Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.

### Testing Instructions
Use the plugin in all ways you can think of.

Use the remember me function of the login.

### Expected result
Nothing changed.

### CODE REVIEW

File: plugins/system/remember/remember.php [hunk 1]
```diff
@@ -131,10 +131,11 @@ public function onUserBeforeSave($user, $isnew, $data)
 		 * But now, we need to do something
 		 * Delete all tokens for this user!
 		 */
-		$db = Factory::getDbo();
+		$db = $this->db;
```
Comment (SharkyKZ) [inline plugins/system/remember/remember.php]: Declare `$db` class variable.

File: plugins/system/remember/remember.php [hunk 2]
```diff
@@ -131,10 +131,11 @@ public function onUserBeforeSave($user, $isnew, $data)
...
 		$query = $db->getQuery(true)
 			->delete('#__user_keys')
```
Comment (Quy) [inline plugins/system/remember/remember.php]:
```suggestion
			->delete($db->quoteName('#__user_keys'))
```

File: plugins/system/remember/remember.php [hunk 3]
```diff
@@ -131,10 +131,11 @@ public function onUserBeforeSave($user, $isnew, $data)
...
-			->where($db->quoteName('user_id') . ' = ' . $db->quote($user['username']));
+			->where($db->quoteName('user_id') . ' = :userid')
+			->bind(':userid', $user['username']);
```
Comment (sandewt) [inline plugins/system/remember/remember.php:134]: I would not expect the username here, but the user id.

This is not desirable in the context of the GDPR !!!!

Comment (brianteeman) [reply plugins/system/remember/remember.php:134]: How on earth is this anything to do with GDPR

Comment (HLeithner) [reply plugins/system/remember/remember.php:134]: This is only a technology refectoring without funtion changr if you think this is wrong please create a new issue and explain in detail ehy this is wrong. Thx

Comment (sandewt) [reply plugins/system/remember/remember.php:134]: @brianteeman and @HLeithner 

Indeed, this is not the right place.

A short explanation:
In itself it is remarkable that the user_id column is used to store the user name and not the user id! See the user_keys table.

File: plugins/system/remember/remember.php [hunk 4]
```diff
@@ -31,6 +31,14 @@ class PlgSystemRemember extends CMSPlugin
 	 */
 	protected $app;
 
+	/**
+	 * Database object
+	 *
+	 * @var    \Joomla\Database\DatabaseInterface
+	 * @since  4.0
+	 */
+	protected $db;
```
Comment (SharkyKZ) [inline plugins/system/remember/remember.php:40]: If I'm not mistaken, this should be `\Joomla\Database\DatabaseDriver`. At least that's what `Joomla\CMS\Factory::getDbo()` documents.

Comment (HLeithner) [reply plugins/system/remember/remember.php:40]: Normally you want to define an interface and you get a object that implements the functionality of the interface. As you can see in the DatabaseDriver definition it implements the DatabaseInterface Interface.

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on 55c29982c8c2cc01ca24695d73936619c2590108<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25539">issues.joomla.org/tracker/joomla-cms/25539</a>.</sub>

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on 55c29982c8c2cc01ca24695d73936619c2590108<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25539">issues.joomla.org/tracker/joomla-cms/25539</a>.</sub>

**Comment by HLeithner** [comment]
Thanks for the tests!

**Pull request merged**